networkx/algorithms/non_randomness.py

```diff
@@ -12,86 +12,150 @@
+    >>> G2.add_edges_from([(0, 1), (1, 3), (1, 2), (3, 4)])
+    >>> [round(r, 6) for r in nx.non_randomness(G1)]
+    [-1.847759, -5.842437]
+    >>> [round(r, 6) for r in nx.non_randomness(G2)]
+    Traceback (most recent call last):
+     ...
+    ValueError: invalid number of communities for graph with 5 nodes and 4 edges: 2
+
+    This is because the community detection algorithm finds
+    1 community in ``G1`` and 2 communities in ``G2``.
+    This can be resolved by specifying the number of communities `k`:
+
+    >>> [round(r, 6) for r in nx.non_randomness(G2, k=1)]
+    [-1.847759, -5.842437]
+
     Notes
     -----
-    This computes Eq. (4.4) and (4.5) in Ref. [1]_.
-
-    If a weight field is passed, this algorithm will use the eigenvalues
-    of the weighted adjacency matrix to compute Eq. (4.4) and (4.5).
+    If a `weight` field is passed, this algorithm will use the eigenvalues
+    of the weighted adjacency matrix instead.
+
+    The output of this function corresponds to (4.4) and (4.5) in [1]_.
+    A lower value of $R^*_G$ indicates a more random graph;
+    one can think of $1 - \\Phi(R_G^*)$ as the similarity
+    between the graph and a random graph,
+    where $\\Phi(x)$ is the cumulative distribution function
+    of the standard normal distribution.
+
+    Theorem 2 in [2]_ states that for any graph $G$
+    with $n$ nodes, $m$ edges, and $k$ communities,
+    its non-randomness is bounded below by the non-randomness of an
+    $r$-regular graph (a graph where each node has degree $r$),
+    and bounded above by the non-randomness of an $l$-complete graph
+    (a graph where each community is a clique of $l$ nodes).
 
     References
     ----------
     .. [1] Xiaowei Ying and Xintao Wu,
            On Randomness Measures for Social Networks,
            SIAM International Conference on Data Mining. 2009
+           https://doi.org/10.1137/1.9781611972795.61
+    .. [2] Ying, Xiaowei & Wu, Leting & Wu, Xintao. (2012).
+           A Spectrum-Based Framework for Quantifying Randomness of Social Networks.
+           IEEE Transactions on Knowledge and Data Engineering 23(12):1842--1856.
+           https://dl.acm.org/doi/abs/10.1109/TKDE.2010.218
     """
     import numpy as np
 
     # corner case: graph has no edges
     if nx.is_empty(G):
-        raise nx.NetworkXError("non_randomness not applicable to empty graphs")
+        raise nx.NetworkXError("graph has no edges")
     if not nx.is_connected(G):
-        raise nx.NetworkXException("Non connected graph.")
+        raise nx.NetworkXException("graph is not connected")
     if len(list(nx.selfloop_edges(G))) > 0:
-        raise nx.NetworkXError("Graph must not contain self-loops")
+        raise nx.NetworkXError("graph contains self-loops")
+
+    n = G.number_of_nodes()
+    m = G.number_of_edges()
 
     if k is None:
         k = len(tuple(nx.community.label_propagation_communities(G)))
+    elif not 1 <= k < n:
```

Comment:
Should this just be an `if`? If `k` >> `n` this approach doesn't really apply, correct? Might as well fail early on user-provided values too.

Reply:
I'm not sure I understand your reasoning. Making it an `if` just means the value of `k` generated by the community detection gets checked as well, but I made the assumption it won't return less than 1 or more than `n-1` communities and so it wouldn't ever catch anything. I guess it could be that the returned number of communities is exactly `n` though? I'm not familiar enough to say for sure.

Reply:
I've updated it to just be an `if`, that way the value set by `label_propagation_communities` gets checked as well.